### PR TITLE
Rejudge

### PR DESCRIPTION
**Pull request by Changron**
1. add rejudge function 

Usage:
To rejudge a submission: `rejudge(submission)`
To rejudge a problem: `rejudge(problem)`
To rejudge a contest: `rejudge(contest)`
To rejudge a problem in contest: `rejudge_contest_problem(contest, problem)`

1. fix my contest showing upcomming contest bug

### CODE REVIEW

**Comment by CHIAHSINHOU** [comment]
Is there anything I can test?

File: contest/templates/contest/contestInfo.html [hunk 1]
```diff
@@ -52,16 +53,20 @@ <h3 class="modal-title">{{ contest.cname }}</h3>
             {% endfor %}
           </tbody>
         </table>
+        {% if user|has_auth:contest.id and contest.problem.count %}
+        <form method="POST" action="{% url 'contest:rejudge' %}">
+          {% csrf_token %}
+          <input type="hidden" name="contest" value="{{ contest.id}}">
+          <button type="submit button" class="btn btn-warning btn-block">
```
Comment (bruce3557) [inline contest/templates/contest/contestInfo.html]: Could we set up alert dialog when rejudge all submission of a problem or all problems?

**Comment by bruce3557** [comment]
Do you check permission to rejudge?

**Comment by nickelcheng** [comment]
@CHIAHSINHOU please test contest rejudge button

File: contest/views.py [hunk 2]
```diff
@@ -384,3 +387,42 @@ def download(request):
             return render_index(request,'contest/download.html',{'contest':contest})
         else:
             raise PermissionDenied
+
+@login_required
+def rejudge(request):
```
Comment (nickelcheng) [inline contest/views.py]: It's required to check permission in this function.
Only contest owner / coowner can rejudge contest problem.

**Comment by Changron** [comment]
add permission check

File: utils/rejudge.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+#rejudge submissions of problem
+def rejudge_problem(problem):
+    submissions = Submission.objects.filter(problem = problem)
+    for submission in submissions:
+        rejudge_submission(submission)
+
+#rejudge single submission
+def rejudge_submission(submission):
+    submission.status = Submission.WAIT
```
Comment (nickelcheng) [inline utils/rejudge.py]: If the original status is ```AC```, submitted problem's ac count should be decreased by 1.

File: utils/rejudge.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    submission.save()
+    logger.info('Submission %s rejudged!' % submission.id)
+    sd = SubmissionDetail.objects.filter(sid = submission)
```
Comment (nickelcheng) [inline utils/rejudge.py]: I prefer to use ```submission_details```

File: utils/rejudge.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    sd_list = SubmissionDetail.objects.filter(sid = submission).\
+              values_list('id', flat=True)
+    sd.delete()
+    for sd_value in sd_list:
```
Comment (nickelcheng) [inline utils/rejudge.py]: Why ```sd_list``` is needed?
Is there any problem to use
```
submission_details = SubmissionDetail.objects.filter(sid = submission)
for submission_detail in submission_details: 
    logger.info('SubmissionDetail %s deleted!' % submission_detail)
submission_details.delete()
```
BTW, I prefer to use the string returned by submission detail object,
which contains "```submission id```: ```testcase id```"

File: utils/rejudge.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        rejudge_contest(obj)
+
+#rejudge submissions of problem
+def rejudge_problem(problem):
+    submissions = Submission.objects.filter(problem = problem)
+    for submission in submissions:
+        rejudge_submission(submission)
+
+#rejudge single submission
+def rejudge_submission(submission):
```
Comment (nickelcheng) [inline utils/rejudge.py]: Please send a message to the submission user, to tell him / her that the submission is rejudged.
About sending message, please use @drowsy810301's code in #197

File: utils/rejudge.py (hunk 5 above)
Comment (Changron) [inline utils/rejudge.py]: If the deletion is not success, the log will still be printed in that way.

**Comment by Changron** [comment]
add that after #197 be merged

File: utils/rejudge.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    if submission.status == Submission.ACCEPTED:
+        submission.problem.ac_count -= 1
+        submission.problem.save()
+    submission.status = Submission.WAIT
+    submission.save()
+    logger.info('Submission %s rejudged!' % submission.id)
+    submission_details = SubmissionDetail.objects.filter(sid = submission)
+    for submission_detail in submission_details:
+        logger.info('SubmissionDetail %s deleted!' % submission_detail)
+        submission_details.delete()
```
Comment (nickelcheng) [inline utils/rejudge.py]: I think this line should not be in the for loop

**Comment by Changron** [comment]
notification added

File: contest/static/contest/js/contestArchive.js [hunk 8]
```diff
@@ -42,6 +42,10 @@ $(document).ready(function() {
     });
 });
 
+function confirm_rejudge() {
+    return confirm("Are you sure you want to rejudge?");
```
Comment (nickelcheng) [inline contest/static/contest/js/contestArchive.js]: Can you ask user to enter total number of submissions need to rejudge? (Of course you have to tell the user first)
Just like when you want to delete a git repository, you need to enter your repository name.

**Comment by nickelcheng** [comment]
+1 besides the comment, @CHIAHSINHOU please test.

**Comment by Changron** [comment]
fixed

**Comment by CHIAHSINHOU** [comment]
This PR is ok for me.

**Comment by nickelcheng** [comment]
going to merge.
@henryyang42  and @geniusgordon you can start your job to rejudge submission / problem.

**Pull request merged**